CAS: Switch from SHA1 to BLAKE3

dexonsmith opened the pull request:
Really we should be adding an enum to choose, but that's kind of hard.
For now, I've reduced the places that name the hash down to two -- a
typedef and BuiltinCAS::getHashName() -- and added some FIXMEs to both
locations that sketch what it could look like.

Why is it hard? Mainly because we'd need to deal with error paths, such
as opening a valid builtin CAS that had a different hash function than
the new/current configuration. Not insurmountable, but let's just move
over.

The downside is that landing this will require a "flag day". Compilers
with this patch can't co-exist with compilers that lack this.

This PR also includes a prep commit "CAS: Remove unused includes of llvm/Support/SHA1.h": cdef20fb1629743b983889de32ea8a35e6dee2b9.

dexonsmith committed:
CAS: Remove unused includes of llvm/Support/SHA1.h

dexonsmith committed:
CAS: Switch from SHA1 to BLAKE3

Switch from SHA1 to BLAKE3 for the builtin CAS.

Really we should be adding an enum to choose, but that's kind of hard.
For now, I've reduced the places that name the hash down to two -- a
typedef and BuiltinCAS::getHashName() -- and added some FIXMEs to both
locations that sketch what it could look like.

Why is it hard? Mainly because we'd need to deal with error paths, such
as opening a valid builtin CAS that had a different hash function than
the new/current configuration. Not insurmountable, but let's just move
over.

Why is it important? I left a big FIXME in BuiltinCAS.h, but the short
version is: future-proofing (for moving to a new hash function) and
storage optimization (using truncated hashes).

This bumps the FilePrefix for the OnDiskCAS from "v1." to "v2." to allow
SHA1 and BLAKE3 objects to coexist at the same CAS path. Without this
change, I can't build locally! This is because my host toolchain is
CAS-enabled (and uses SHA1), and I have `-fdepscan` on for both the host
clang and for tablegen (which uses BLAKE3).

dexonsmith commented:
I don't even know if this compiles, since I'm having trouble with my local toolchain, but I think the two patches are correct?

@swift-ci test

dexonsmith commented:
> I don't even know if this compiles, since I'm having trouble with my local toolchain, but I think the two patches are correct?

It does not compile due to https://reviews.llvm.org/D121510#inline-1174446 — `BLAKE3::final()` works differently from all the other hashers.

benlangmuir commented on llvm/lib/CAS/BuiltinCAS.h:
```diff
@@ -107,15 +107,38 @@ inline ArrayRef<char> toArrayRef(StringRef Data) {
 /// Current hash type for the internal CAS.
-using HasherT = SHA1;
+///
+/// FIXME: This should be configurable via an enum, with at the following values:
```
What's the rationale for making it configurable? What situations would demand 16-byte or 32-byte hashes, respectively?

dexonsmith replied:
Making it configurable is important for:
- Future proofing, to allow us to eventually move to BLAKE7.
- Optimizing memory usage and storage. My reading of the BLAKE3 paper (which could be faulty) gave me the impression that 16B would be a good default for the in-memory CAS and likely many local use cases... seemed to imply it was already as good as or better than SHA-1 (which is what Git uses). Maybe we want an in-memory-only CAS to default to 16B; maybe we want everything to default to 16B; maybe a vendor choice, or a use-case choice made by the build system (using this on-disk CAS for a single build folder? 16B. Using this one for all clang/LLVM on the machine? 24B. Hooking it up to a massive distributed system? 32B).

dexonsmith replied:
Two follow-ups, depending on your response:
- If you agree, I'll add the rationale to the commit.
- Maybe we should use 16B or 20B now — WDYT?

benlangmuir replied:
> Making it configurable is important for:
> * Future proofing, to allow us to eventually move to BLAKE7.

Ah sorry, I should have been more specific that I was responding more to the idea of configuring between blake3 and blake3_16, since those are the cases you listed.  I agree that we would want to be able to handle a new better hash in the future, but it isn't obvious to me that we would actually want to add a _parameter_ for that until we have such a hash function in hand, since the API is unstable anyway.

> Optimizing memory usage and storage.

Maybe this is obvious to others, but where do we materialize CASIDs as full hashes rather than using indirection? How important would the saving be?

> My reading of the BLAKE3 paper... Maybe we want an in-memory-only CAS to default to 16B; maybe we want everything to default to 16B; maybe a vendor choice, or a use-case choice made by the build system ...

If there is a significant benefit to using smaller hashes for the in-memory CAS, and it has good enough hash quality for that use case, that seems like a fine reason to make it configurable to me.

> maybe we want everything to default to 16B
> Maybe we should use 16B or 20B now — WDYT?

If we feel that 32B is necessary for some cases, I would keep 32B as the default for the on-disk format for now.

dexonsmith replied:
> > Optimizing memory usage and storage.
> 
> Maybe this is obvious to others, but where do we materialize CASIDs as full hashes rather than using indirection? How important would the saving be?

- They are stored in the CAS itself, one per object. Saving 16B per object if you have 1M objects saves 16MB. I don't know if this matters, but this is what I was referring to previously. Also:
- Any non-integrated action cache needs to store the IDs of the action results that are cached.
- An action cache that lives longer than the CAS also needs to recursively store any objects. If we support adapting a generic hash map (where the object graph is opaque), it won't be able to optimize the CAS references.

The latter two would play out something like this:
```
/// A generic map interface.
class GenericMap {
  /// Insert into the map, calling ValueWriter if and only if Key is new.
  virtual Error put(const Twine &Key, function_ref<Error (raw_ostream &)> ValueWriter) = 0;
  virtual Error read(const Twine &Key, raw_ostream &OS) const = 0;
  virtual Error poison(const Twine &Key) = 0;
};

/// Adapt a generic map to cache actions. Assume result objects are stored elsewhere.
class MapToActionCacheAdaptor : public ActionCache {
  Map &M;
public:
  Error put(const ActionKey &Key, const CASID &Result) override {
    return M.put(Key.toString(), [&]() { return Result.toString()}).takeError();
  }
  // ...
};

/// Adapt a generic map to cache actions. Store result objects.
class MapToActionCacheAndStorageAdaptor : public ActionCache {
  Map &M;
  ObjectStore &CAS;

public:
  Error put(const ActionKey &Key, const CASID &Result) {
    SmallVector<CASID> Worklist = {Result};
    DenseMap<CASID> Visited;
    auto addChildToWorklist = [&](const CASID &ID) {
      if (Visited.insert(CASID).second)
        Worklist.push_back(Child);
    };

    while (!Worklist.empty()) {
      CASID ID = Worklist.pop_back_val();
      auto serializeObject = [&](raw_ostream &OS) -> Error {
        return CAS.serializeObject(OS, ID, addChildToWorklist));
      };

      // If the object is already there, serializeObject won't be called, and we
      // don't need to descend.
      if (Error E = M.put(StringRef("object:") + Result.toString(), serializeObject).moveInto(WasInserted))
        return E;
    }
    return M.put(StringRef("action:") + Key.toString(), Result.toString());
  }
  // ...
};
```
The `serializeObject` for a node would be something like:
```
endian::writer EW(OS, endian::little);
EW.write<uint32_t>(Obj.getNumRefs());
EW.write<uint32_t>(CAS.getHashSize());
for (const CASID &Child : Obj.children()) {
  EW << Child.toString();
  addChildToWorklist(Child);
}
EW.write<uint64_t>(CAS.getData().size());
EW.write(Obj.getData());
```
So each CAS reference here ends up costing the full amount in the object that references it.

... That said, maybe this says more about `MapToActionCacheAndStorageAdaptor`'s suitability as a plugin interface than anything else.

dexonsmith committed:
CAS: Add a test dedicated to printing/parsing IDs

A future commit exposes a bug in OnDiskHashMappedTrie / OnDiskCAS that
comes up when printing IDs. Add a test dedicated to printing IDs so that
future bugs are easier to debug.

dexonsmith committed:
CAS: Fix recoverFromFileOffset() for 8B-aligned hashes

Fix an off-by-one error in
OnDiskHashMappedTrie::recoverFromFileOffset(). This never hits for SHA1,
because it's not 8B-aligned, so there's always padding between the last
insertion and the end of the file.

It'll be exposed by switching to BLAKE3.

dexonsmith commented:
@swift-ci test

dexonsmith commented:
Just force-pushed an update.
- 63f94429c893119e48dae990f6f949360b1a2952 fixes a bug in OnDiskHashMappedTrie that BLAKE3 uncovered since it happens to be 8B-aligned (and SHA1 is not).
- 32d0d84eba5fdf64500d2b9128c91d96cb07c86f adds a test in CASDB dedicated to printing/parsing IDs, which I wrote to help track down the bug (catches it at a higher level), but I think is independently useful.
- d57cbf80f8b69e9ac6172aa3ffbdc79a43357406 removes unused includes of llvm/Support/SHA1.h, and adds one to HashMappedTrieTest.cpp since that's using SHA1 directly in the tests.
- 6ddc4144461c0d1571069d79561e733b7311a9f9 is the updated main commit. I expanded the FIXME substantially based on @benlangmuir's comments / questions. I also changed the FilePrefix from `v1.` to `v2.` since otherwise I couldn't build locally (explanation in the commit message)... as a result, we no longer need a flag day.

PTAL!

dexonsmith commented on llvm/unittests/CAS/OnDiskHashMappedTrieTest.cpp:
```diff
@@ -0,0 +1,141 @@
+    EXPECT_FALSE(Trie1->find(Hash1));
+    EXPECT_FALSE(Trie2->find(Hash1));
+
+    // Recover from an offset.
+    {
+      auto Recovered = Trie1->recoverFromFileOffset(*Offset);
+      ASSERT_TRUE(Recovered);
+      EXPECT_EQ(Offset->get(), Recovered.getOffset().get());
+      EXPECT_EQ(Hash0, Recovered->Hash);
+      EXPECT_EQ(Data0v2, Recovered->Data);
+    }
```
FYI, this is the "payload", so to speak, of the test (this is what failed without the fix to OnDiskHashMappedTrie.cpp at https://github.com/apple/llvm-project/pull/4149/files#diff-d5b3ea2752ca6dc3ca7552592c549b3b43bc07a3c269e542857fbf188906141bR663). The rest of this was just necessary since there were no unit tests for OnDiskHashMappedTrie until now :/.

benlangmuir commented on llvm/lib/CAS/OnDiskCAS.cpp:
```diff
@@ -644,7 +644,7 @@ struct OnDiskObjectProxy {
 ///
 /// Files:
-/// - db/v1.index: HashMappedTrie(name="cas.hashes[sha1]")
+/// - db/v1.index: HashMappedTrie(name="cas.hashes[<hashname>]")
```
Nitpick: Since the comment above says this is the "current layout", should we bump the file prefix here to match reality, or make it `v<N>` or something?

dexonsmith replied:
Actually, the whole thing is kind of wrong at this point; that was documentation from before I'd finished the branch that rewrote the CAS. I'll take a look and either update it or delete the parts that seem likely to bitrot (maybe leave behind pointers for where to look to determine the layout, or move the comment somewhere it'll be updated).

dexonsmith replied:
Fixed in 38288e7fb076b5722581c7944f16889d5d336441.

benlangmuir commented on llvm/lib/CAS/OnDiskHashMappedTrie.cpp:
```diff
@@ -660,7 +660,7 @@ OnDiskHashMappedTrie::recoverFromFileOffset(FileOffset Offset) const {
   const uint64_t CurrentFileSize = Impl->File.getAlloc().size();
   if (Offset.get() > (int64_t)CurrentFileSize ||
-      Offset.get() + Impl->Trie.getRecordSize() >= CurrentFileSize)
+      Offset.get() + Impl->Trie.getRecordSize() > CurrentFileSize)
```
Nitpick: `Impl->Trie.getRecordSize() > CurrentFileSize - Offset.get()` avoids a potential overflow, although if we overflow an `int64_t` it probably means something has gone horribly wrong already

dexonsmith replied:
Good idea. Horrible things have happened before; might as well avoid the potential overflow :).

dexonsmith replied:
Fixed in ce1d1a22853bcdffc49f127fe222c54ce827d49c.

benlangmuir reviewed (approved):
Couple of minor comments, but LGTM. 

One more clarification: you mentioned
> This never hits for SHA1, because it's not 8B-aligned

Is it really the alignment of the hash that matters, or is it because the stride/size of BLAKE3's hash is a multiple of 8?  I would have expected both hashes to have an alignment of 1 since you're representing them as an array of bytes.

dexonsmith commented:
> Couple of minor comments, but LGTM.
> 
> One more clarification: you mentioned
> 
> > This never hits for SHA1, because it's not 8B-aligned
> 
> Is it really the alignment of the hash that matters, or is it because the stride/size of BLAKE3's hash is a multiple of 8? I would have expected both hashes to have an alignment of 1 since you're representing them as an array of bytes.

You're correct. "Aligned" was not the right term, I meant a multiple of 8.

dexonsmith committed:
CAS: Avoid potential overflow in recoverFromFileOffset()

Avoid a potential (but exceedingly unlikely!) overflow
noticed when reviewing 63f94429c893119e48d.

dexonsmith committed:
CAS: Update description for OnDiskCAS layout

Update the description for the layout of the OnDiskCAS on disk. It was
originally written before being implemented, and had bitrotted as
files and tables were renamed. The new description:

- Refers to helper functions for table names, so that they're less
  likely to bitrot (since when the helper functions are renamed a local
  file search should catch this too).
- Describes missing pieces that didn't exist earlier (such as the
  actions table and standalone blobs).
- More precisely describes the contents, pointing at other helper
  data structures.

In response to review of 6ddc4144461c0d1571069d79561e733b7311a9f9, which
had lazily made the description less wrong without bothering to actually
fix it.

dexonsmith committed:
CAS: Clarify 8B-aligned language in OnDiskHashMappedTrieTest, NFC

Update the comments about "8B-aligned" an data / hash added in
6ddc4144461c0d1571069d79561e733b7311a9f9 for OnDiskHashMappedTrieTest
after review.

- I didn't mean "8B-aligned", I meant "multiples of 8B".
- Explain why it matters!

dexonsmith commented:
> > Couple of minor comments, but LGTM.
> > One more clarification: you mentioned
> > > This never hits for SHA1, because it's not 8B-aligned
> > 
> > 
> > Is it really the alignment of the hash that matters, or is it because the stride/size of BLAKE3's hash is a multiple of 8? I would have expected both hashes to have an alignment of 1 since you're representing them as an array of bytes.
> 
> You're correct. "Aligned" was not the right term, I meant a multiple of 8.

Updated the test comments in ebd970947f03fae0393126326aa64c3c59cfe1b1.

dexonsmith commented:
(thanks for the review @benlangmuir !)
